src/marshmallow/schema.py

```diff
@@ -1132,33 +1133,22 @@ def _invoke_processors(
         data,
         many,
         original_data=None,
+        **kwargs
     ):
         key = (tag, pass_many)
         for attr_name in self._hooks[key]:
             # This will be a bound method.
             processor = getattr(self, attr_name)
 
-            processor_kwargs = processor.__marshmallow_hook__[key]
-            pass_original = processor_kwargs.get('pass_original', False)
-
             if pass_many:
-                if pass_original:
-                    data = processor(data, many, original_data)
-                else:
-                    data = processor(data, many)
+                data = processor(data, many=many, original_data=original_data, **kwargs)
             elif many:
-                if pass_original:
-                    data = [
-                        processor(item, original)
-                        for item, original in zip(data, original_data)
-                    ]
-                else:
-                    data = [processor(item) for item in data]
+                data = [
+                    processor(item, original_data=original, **kwargs)
+                    for item, original in zip_longest(data, original_data)
```

Comment:
I'm having second thoughts about dropping the `pass_original` parameter because it forces us to zip the processed data with `original_data` even if the user doesn't need the original data, which incurs a performance penalty for both dump-ing and load-ing.

Reply:
Even worse, this behavior can be incorrect when `original_data` doesn't have the same shape as  `data` (e.g. when `original_data` is enveloped).

Reply:
Reminds me of https://github.com/marshmallow-code/marshmallow/issues/1150.

The case where `original_data` is not iterable is not totally covered.